src/parsers/doctype.js

```diff
@@ -0,0 +1,45 @@
+const ENTITY_RE = /<!ENTITY\s+([\w.]*)\s+("[^"]*"|'[^']*')\s*>/;
+
+
+export default class DoctypeParser {
+  /*
+   * Minimalistic parser for DTD files.
+   * We are using DTD files in our language packs to specify translations.
+   *
+   * It doesn't do any proper XML parsing because our DTD files are only
+   * using a very small subset of what's needed so a simple regular expression
+   * works just fine.
+   *
+   * The parsing code is largely inspired by scattered code from mozilla-central.
+   */
+
+  constructor(dtdString, collector, { filename = null } = {}) {
+    this._dtdString = dtdString;
+    this.collector = collector;
+    this.filename = filename;
+    this.isValid = null;
+  }
+
+  parse() {
+    const entities = this._dtdString.match(new RegExp(ENTITY_RE, 'gi'));
+    this.parsedData = {};
+
+    if (!entities) {
+      // Some files have no entities defined. Mark this as valid since we
+      // expect this unfortunately...
+      this.isValid = true;
+      return;
+    }
+
+    entities.forEach((entity) => {
+      const [, key, value] = entity.match(new RegExp(ENTITY_RE, 'i'));
+
+      // strip enclosing quotation marks
+      const normalizedValue = value.slice(1, -1);
+
+      this.parsedData[key] = normalizedValue;
+    });
+
+    this.isValid = true;
```

Comment:
It looks like this always sets `this.isValid = true`. Would this throw on an invalid dtd file?

Reply:
If the file is completely invalid, yes. We parse dtd files very resilient to errors though so that we don't care right now. We mostly just care about whether or not we can open the file. More checking might come later.